### ISSUE TITLE
Can't launch Vulkan App when using `VkPhysicalDeviceBufferDeviceAddressFeatures` in the `VkDeviceCreateInfo` next pointer chain

### ISSUE THREAD

**Issue by eyerust** [baldurk/renderdoc#2059]
## Description

I want to launch a vulkano (https://github.com/vulkano-rs/vulkano/) app using RenderDoc but it fails to create a vulkan device. 
Here is the log of trying to run the vulkano triangle example: [renderdoc-triangle.log](https://github.com/baldurk/renderdoc/files/5264012/renderdoc-triangle.log)
But the code can be reduced to just the instance and device creation
```
use std::sync::Arc;
use vulkano::device::Features;
use vulkano::device::{Device, DeviceExtensions};
use vulkano::instance::InstanceExtensions;
use vulkano::instance::{Instance, PhysicalDevice};

fn main() {
    let instance = Instance::new(None, &InstanceExtensions::none(), None).unwrap();
    let physical = PhysicalDevice::enumerate(&instance).next().unwrap();

    let queue_family = physical.queue_families().next().unwrap();

    let (device, mut queues) = Device::new(
        physical,
        &Features::none(),
        &DeviceExtensions::none(),
        [].iter().cloned(),
    )
    .unwrap();
}
```
The hello_triangle vulkan sample runs without problems so I assumed that the problem lies within Vulkano. But after some investigation it seems to be a problem within RenderDoc.
I already filed an issue to Vulkano (https://github.com/vulkano-rs/vulkano/issues/1416). There is more information about my problem.
Vulkano sets the `pNext` pointer instead of the `pEnabledFeatures` pointer and uses physical device feature extensions. This seems to be the only reason why it doesn't work but the VulkanSamples work.
Here is the log when setting `pNext` (and `pEnabledFeatures`) to null: [renderdoc-fixed.log](https://github.com/baldurk/renderdoc/files/5264045/renderdoc-fixed.log)
The device creation succeeds but this doesn't help me because I (or Vulkano) need the `pNext` pointer.

I also found two related issues: #1744 and #1157 which seem to describe my problem (or at least something similar) and are closed but they seem to not apply to my situation.

### So in a nutshell:
*What I want:* Use `pNext` pointer in `DeviceCreatInfo` to point to `VkPhysicalDeviceFeatures2` or similar and launch using RenderDoc.
*What happens:* App crashes because device creation failed. Happens only with RenderDoc.
*What I expected to happen:* Just work...

## Steps to reproduce

* Set `pNext` of `DeviceCreateInfo` to a valid value in any vulkan app or use the example Rust code from above
* Launch the app with RenderDoc
* See that the device creation failed

## Environment

* RenderDoc version: 1.10
* Operating System: Gentoo Linux amd64
* Graphics API: Vulkan 1.1.0

**Comment by baldurk** [comment]
I do not believe it is accurate to say that any valid value of `vkDeviceCreateInfo::pNext` will crash. What is possible is that some particular structs on some particular drivers can fail, but I'd need more information to narrow down what is happening. I will try to run the Rust example but since Rust is painful to set up I might be able to see the problem more quickly if you can tell me exactly what you are passing to the `vkCreateDevice` call.

Just to sanity check, is your code correctly checking for all features that you enable before you enable them? The sample you posted doesn't specify any of this so I assume it's happening in your vulkan wrapper library.

I'll look into this further but it won't be until later next week as I'm off work just now. However I will need more explicit instructions on how to reproduce the problem using your rust code since I don't know how to compile a file like that into an executable I can run.

**Comment by eyerust** [comment]
Ok, I managed to narrow it down to a simple C++ example.
```
#include <vulkan/vulkan.h>
#include <iostream>
#include <stdexcept>
#include <vector>
#include <cstring>
#include <cstdlib>
#include <optional>

const std::vector<const char*> validationLayers = {
    "VK_LAYER_KHRONOS_validation",
    "VK_LAYER_RENDERDOC_Capture"
};

int main() {
    VkApplicationInfo appInfo{};
    appInfo.sType = VK_STRUCTURE_TYPE_APPLICATION_INFO;
    appInfo.pApplicationName = "Hello Triangle";
    appInfo.applicationVersion = VK_MAKE_VERSION(1, 0, 0);
    appInfo.pEngineName = "No Engine";
    appInfo.engineVersion = VK_MAKE_VERSION(1, 0, 0);
    appInfo.apiVersion = VK_API_VERSION_1_1;

    VkInstanceCreateInfo createInfo{};
    createInfo.sType = VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO;
    createInfo.pNext = nullptr;
    createInfo.flags = 0;
    createInfo.pApplicationInfo = &appInfo;
    createInfo.ppEnabledLayerNames = validationLayers.data();
    createInfo.enabledLayerCount = validationLayers.size();

    VkInstance instance = nullptr;

    if (vkCreateInstance(&createInfo, nullptr, &instance) != VK_SUCCESS) {
        throw std::runtime_error("failed to create instance!");
    }

    uint32_t deviceCount = 0;
    vkEnumeratePhysicalDevices(instance, &deviceCount, nullptr);

    std::vector<VkPhysicalDevice> devices(deviceCount);
    vkEnumeratePhysicalDevices(instance, &deviceCount, devices.data());

    VkPhysicalDevice physicalDevice = devices.front();

    VkDeviceQueueCreateInfo queueCreateInfo{};
    queueCreateInfo.sType = VK_STRUCTURE_TYPE_DEVICE_QUEUE_CREATE_INFO;
    queueCreateInfo.queueFamilyIndex = 0;
    queueCreateInfo.queueCount = 1;
    float queuePriority = 1.0f;
    queueCreateInfo.pQueuePriorities = &queuePriority;

    VkPhysicalDeviceFeatures features{};
    features.robustBufferAccess = true;

    // How it is officially supported
    VkPhysicalDeviceBufferDeviceAddressFeatures moreFeatures{};
    moreFeatures.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_BUFFER_DEVICE_ADDRESS_FEATURES;
    moreFeatures.pNext = nullptr;

    // How Vulkano does it
    // VkPhysicalDeviceBufferAddressFeaturesEXT moreFeatures{};
    // moreFeatures.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_BUFFER_ADDRESS_FEATURES_EXT;
    // moreFeatures.pNext = nullptr;

    // They should do the same.
    // Both work without RenderDoc. None with RenderDoc.

    VkPhysicalDeviceFeatures2 features2{};
    features2.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_FEATURES_2;
    features2.pNext = &moreFeatures;
    features2.features = features;

    VkDeviceCreateInfo deviceCreateInfo{};
    deviceCreateInfo.sType = VK_STRUCTURE_TYPE_DEVICE_CREATE_INFO;
    deviceCreateInfo.pNext = &features2;
    deviceCreateInfo.pQueueCreateInfos = &queueCreateInfo;
    deviceCreateInfo.queueCreateInfoCount = 1;
    deviceCreateInfo.pEnabledFeatures = nullptr;
    deviceCreateInfo.ppEnabledLayerNames = validationLayers.data();
    deviceCreateInfo.enabledLayerCount = validationLayers.size();

    VkDevice device = nullptr;

    if (vkCreateDevice(physicalDevice, &deviceCreateInfo, nullptr, &device) != VK_SUCCESS) {
        throw std::runtime_error("failed to create logical device!");
    }

    vkDestroyDevice(device, nullptr);
    vkDestroyInstance(instance, nullptr);
}
```
This does nothing more than to create a device with additional device features and validation layers enabled.
Vulkano seems to use `VkPhysicalDeviceBufferAddressFeaturesEXT` but I can't find it in the official specs so I used `VkPhysicalDeviceBufferDeviceAddressFeatures` for this example which seems to do the same thing and is officially supported if I didn't misread the specs.
I don't enable any features but just specify `VkPhysicalDeviceBufferDeviceAddressFeatures` in the next chain. Without it, everything works.
* GPU: NVIDIA GeForce GTX 1060 6GB
* GPU Drivers: NVIDIA Accelerated Graphics Driver (450.66)

**Comment by baldurk** [comment]
That code looks invalid to me. You must not pass a struct in a `pNext` chain unless you have enabled the requisite core version or extensions. `VkPhysicalDeviceBufferDeviceAddressFeatures` or its aliases either come from either `KHR/EXT_buffer_device_address` or vulkan core 1.2. You have enabled vulkan core 1.1 with no extensions, and so you must not pass that struct. Does it still happen if you check to see that the extension is available and then enable it?

Note also that you should not ever enable RenderDoc's layer manually like that. That is not designed to work and is not supported, you should always allow the RenderDoc UI to enable the layer for you.

**Comment by eyerust** [comment]
Okay I changed the code to this
```
#include <vulkan/vulkan.h>
#include <iostream>
#include <stdexcept>
#include <vector>
#include <cstring>
#include <cstdlib>
#include <optional>

const std::vector<const char*> validationLayers = {
    "VK_LAYER_KHRONOS_validation"
};

int main() {
    VkApplicationInfo appInfo{};
    appInfo.sType = VK_STRUCTURE_TYPE_APPLICATION_INFO;
    appInfo.pApplicationName = "Hello Triangle";
    appInfo.applicationVersion = VK_MAKE_VERSION(1, 0, 0);
    appInfo.pEngineName = "No Engine";
    appInfo.engineVersion = VK_MAKE_VERSION(1, 0, 0);
    appInfo.apiVersion = VK_API_VERSION_1_1;

    const std::vector<const char*> instanceExtensions = {
        "VK_KHR_get_physical_device_properties2"
    };

    VkInstanceCreateInfo createInfo{};
    createInfo.sType = VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO;
    createInfo.pNext = nullptr;
    createInfo.flags = 0;
    createInfo.pApplicationInfo = &appInfo;
    createInfo.ppEnabledLayerNames = validationLayers.data();
    createInfo.enabledLayerCount = validationLayers.size();

    createInfo.enabledExtensionCount = instanceExtensions.size();
    createInfo.ppEnabledExtensionNames = instanceExtensions.data();

    uint32_t supportedInstanceExtensionCount = 0;
    vkEnumerateInstanceExtensionProperties(nullptr, &supportedInstanceExtensionCount, nullptr);
    std::vector<VkExtensionProperties> supportedInstanceExtensions(supportedInstanceExtensionCount);
    vkEnumerateInstanceExtensionProperties(nullptr, &supportedInstanceExtensionCount, supportedInstanceExtensions.data());

    std::cout << "available instance extensions:\n";

    for (const auto& extension : supportedInstanceExtensions) {
        std::cout << '\t' << extension.extensionName << '\n';
    }

    VkInstance instance = nullptr;

    if (vkCreateInstance(&createInfo, nullptr, &instance) != VK_SUCCESS) {
        throw std::runtime_error("failed to create instance!");
    }

    uint32_t deviceCount = 0;
    vkEnumeratePhysicalDevices(instance, &deviceCount, nullptr);

    std::vector<VkPhysicalDevice> devices(deviceCount);
    vkEnumeratePhysicalDevices(instance, &deviceCount, devices.data());

    VkPhysicalDevice physicalDevice = devices.front();

    VkDeviceQueueCreateInfo queueCreateInfo{};
    queueCreateInfo.sType = VK_STRUCTURE_TYPE_DEVICE_QUEUE_CREATE_INFO;
    queueCreateInfo.queueFamilyIndex = 0;
    queueCreateInfo.queueCount = 1;
    float queuePriority = 1.0f;
    queueCreateInfo.pQueuePriorities = &queuePriority;

    VkPhysicalDeviceFeatures features{};
    features.robustBufferAccess = true;

    // How it is officially supported
    VkPhysicalDeviceBufferDeviceAddressFeatures moreFeatures{};
    moreFeatures.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_BUFFER_DEVICE_ADDRESS_FEATURES;
    moreFeatures.pNext = nullptr;

    // How Vulkano does it
    // VkPhysicalDeviceBufferAddressFeaturesEXT moreFeatures{};
    // moreFeatures.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_BUFFER_ADDRESS_FEATURES_EXT;
    // moreFeatures.pNext = nullptr;

    // They should do the same.
    // Both work without RenderDoc. None with RenderDoc.

    VkPhysicalDeviceFeatures2 features2{};
    features2.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_FEATURES_2;
    features2.pNext = &moreFeatures;
    features2.features = features;

    VkDeviceCreateInfo deviceCreateInfo{};
    deviceCreateInfo.sType = VK_STRUCTURE_TYPE_DEVICE_CREATE_INFO;
    deviceCreateInfo.pNext = &features2;
    deviceCreateInfo.pQueueCreateInfos = &queueCreateInfo;
    deviceCreateInfo.queueCreateInfoCount = 1;
    deviceCreateInfo.pEnabledFeatures = nullptr;
    deviceCreateInfo.ppEnabledLayerNames = validationLayers.data();
    deviceCreateInfo.enabledLayerCount = validationLayers.size();

    const std::vector<const char*> deviceExtensions = {
        "VK_KHR_buffer_device_address"
    };

    deviceCreateInfo.enabledExtensionCount = deviceExtensions.size();
    deviceCreateInfo.ppEnabledExtensionNames = deviceExtensions.data();

    uint32_t supportedDeviceExtensionCount = 0;
    vkEnumerateDeviceExtensionProperties(physicalDevice, nullptr, &supportedDeviceExtensionCount, nullptr);
    std::vector<VkExtensionProperties> supportedDeviceExtensions(supportedDeviceExtensionCount);
    vkEnumerateDeviceExtensionProperties(physicalDevice, nullptr, &supportedDeviceExtensionCount, supportedDeviceExtensions.data());

    std::cout << "available device extensions:\n";

    for (const auto& extension : supportedDeviceExtensions) {
        std::cout << '\t' << extension.extensionName << '\n';
    }

    VkDevice device = nullptr;

    if (vkCreateDevice(physicalDevice, &deviceCreateInfo, nullptr, &device) != VK_SUCCESS) {
        throw std::runtime_error("failed to create logical device!");
    }

    vkDestroyDevice(device, nullptr);
    vkDestroyInstance(instance, nullptr);
}
```

I enabled `VK_KHR_buffer_device_address` and its dependencies.
I also print a list of supported device extensions which definitely includes `VK_KHR_buffer_device_address`.
When I launch it with RenderDoc I get a new warning: `Core 11512 14:48:44 vk_core.cpp(3729) Warning [Loader Message] loader_validate_device_extensions: Device extension VK_KHR_buffer_device_address not supported by selected physical device or enabled layers.`
I only enabled the layer `VK_LAYER_KHRONOS_validation` and `VK_KHR_buffer_device_address` is supported by my device. So why this warning?
Oh I forgot, here the log: [minimal-log.log](https://github.com/baldurk/renderdoc/files/5268200/minimal-log.log)
Ah and the same happens with `VK_EXT_buffer_device_address`

**Comment by baldurk** [comment]
You don't actually verify that the extension you're enabling is available, you only print the available extensions and then enable `VK_KHR_buffer_device_address` anyway. RenderDoc modifies the list of extensions available to only those ones it can support for capture and replay, which may mean it removes some extensions that are otherwise available. Your application must correctly check for extensions it enables before running.

Judging by the log your GPU and driver doesn't support the `bufferDeviceAddressCaptureReplay` feature meaning RenderDoc can't support the buffer device address feature at all. If you have a proper extension and feature check then you will be able to detect that when running under RenderDoc, then it's up to you whether you can support a fallback path to not use the extension or fail to start.

This seems like an application bug then, not a RenderDoc bug.

**Comment by eyerust** [comment]
Yes this seems to be a problem with Vulkano, not RenderDoc. Thank you very much for your help, now I understand what the problem is. Vulkano *always* uses a feature that depends on an extension (`VK_KHR_buffer_device_address`) that is only missing with RenderDoc. I should return to my original issue...
